File: preprocess/preprocess.py
#!/usr/bin/env python3
"""
Preprocess depth RDS chunks into browser-readable files for GitHub Pages.

Reads the five Chunk_*.rds files (each an R named list of per-site data
frames) plus Depth_Site_Name_Crosswalk.rds, and writes:

  data/manifest.json          - site index, crosswalk, format documentation
  data/sites/<SITE>.json.gz   - one gzipped columnar JSON file per site

Original values are preserved exactly:
  * DateTime           - original character timestamps, verbatim
  * DateTime_GMT_minus_5 - POSIXct epoch seconds, delta-encoded (lossless,
                           round-trip verified below)
  * Depth_m            - full float64 precision (shortest round-trip repr)
  * Phase, Source_Block - run-length encoded against a per-site dictionary

The browser never runs R: it fetches one .json.gz per selected site and
decompresses it with the native DecompressionStream API.

Requires: pip install rdata pyreadr pandas numpy
Usage:    python3 preprocess/preprocess.py   (run from the repository root)
"""

import gzip
import json
import math
import os
import sys
import warnings
from datetime import datetime, timezone

import numpy as np

warnings.filterwarnings("ignore")

CHUNK_FILES = [
    "Chunk_01_Blue_Ridge.rds",
    "Chunk_02_Central_Maryland.rds",
    "Chunk_03_PRV.rds",
    "Chunk_04_Roberts_Field.rds",
    "Chunk_05_Shannon_Run.rds",
]
CROSSWALK_FILE = "Depth_Site_Name_Crosswalk.rds"
EXPECTED_SITES = ["BR_2", "BR_3", "CM_1", "CM_2", "PRV_1", "PRV_2",
                  "RF_1", "RF_2", "SR_1", "SR_3"]
OUT_DIR = "data"
FORMAT_VERSION = 1


def read_chunks():
    """Read every chunk RDS; each is a named list of per-site data frames."""
    import rdata
    sites = {}
    for f in CHUNK_FILES:
        print(f"Reading {f} ...", flush=True)
        obj = rdata.conversion.convert(rdata.parser.parse_file(f))
        if not isinstance(obj, dict):
            sys.exit(f"{f}: expected a named list of data frames, got {type(obj)}")
        for site, df in obj.items():
            site = str(site)
            if site in sites:
                sys.exit(f"Site {site} appears in more than one chunk file")
            sites[site] = (f, df)
    return sites


def read_crosswalk():
    import pyreadr
    df = list(pyreadr.read_r(CROSSWALK_FILE).values())[0]
    rows = []
    for _, r in df.iterrows():
        rows.append({k: (None if (isinstance(v, float) and math.isnan(v)) or v is None
                         else (v if not isinstance(v, float) else v))
                     for k, v in r.items()})
    return rows


def rle_encode(codes):
    """Run-length encode an integer code sequence as [[code, count], ...]."""
    runs = []
    prev, count = None, 0
    for c in codes:
        if c == prev:
            count += 1
        else:
            if prev is not None:
                runs.append([prev, count])
            prev, count = c, 1
    if prev is not None:
        runs.append([prev, count])
    return runs


def dict_rle(series):
    values = series.tolist()
    dictionary = []
    index = {}
    codes = []
    for v in values:
        v = "" if v is None else str(v)
        if v not in index:
            index[v] = len(dictionary)
            dictionary.append(v)
        codes.append(index[v])
    return dictionary, rle_encode(codes)


def encode_time(t):
    """Delta-encode epoch seconds. t[i] is null when missing; otherwise the
    stored value is t[i] minus the previous non-null t (first non-null is
    stored relative to t0). Deltas may be negative (blocks can overlap)."""
    out = []
    t0 = None
    prev = None
    for v in t:
        if v is None or (isinstance(v, float) and math.isnan(v)):
            out.append(None)
            continue
        iv = v
        # POSIXct values here are whole seconds; keep an int when exact.
        if float(iv) == int(iv):
            iv = int(iv)
        if t0 is None:
            t0 = iv
            out.append(0)
        else:
            out.append(iv - prev)
        prev = iv
    return t0, out


def decode_time(t0, deltas):
    out = []
    prev = None
    for d in deltas:
        if d is None:
            out.append(None)
            continue
        cur = t0 if prev is None else prev + d
        out.append(cur)
        prev = cur
    return out


def depth_list(series):
    out = []
    for v in series.to_numpy(dtype=float):
        out.append(None if math.isnan(v) else float(v))
    return out


def main():
    crosswalk = read_crosswalk()
    xwalk_by_site = {r["Final_List_Name"]: r for r in crosswalk}

    sites = read_chunks()
    missing = [s for s in EXPECTED_SITES if s not in sites]
    extra = [s for s in sites if s not in EXPECTED_SITES]
    if missing:
        sys.exit(f"Missing expected sites: {missing}")
    if extra:
        print(f"WARNING: unexpected extra sites present: {extra}")

    os.makedirs(os.path.join(OUT_DIR, "sites"), exist_ok=True)
    manifest_sites = []

    for site in EXPECTED_SITES + extra:
        chunk_file, df = sites[site]
        n = len(df)
        print(f"Encoding {site}: {n} rows (from {chunk_file})", flush=True)

        expected_cols = ["DateTime", "DateTime_GMT_minus_5", "Depth_m",
                         "Phase", "Source_Block"]
        for c in expected_cols:
            if c not in df.columns:
                sys.exit(f"{site}: missing column {c}")

        t_raw = [None if math.isnan(v) else float(v)
                 for v in df["DateTime_GMT_minus_5"].to_numpy(dtype=float)]
        t0, t_delta = encode_time(t_raw)

        # Lossless round-trip check on the time encoding.
        decoded = decode_time(t0, t_delta)
        for i, (a, b) in enumerate(zip(t_raw, decoded)):
            ok = (a is None and b is None) or (a is not None and b is not None
                                               and float(a) == float(b))
            if not ok:
                sys.exit(f"{site}: time round-trip mismatch at row {i}: {a} != {b}")

        dt = ["" if v is None else str(v) for v in df["DateTime"].tolist()]
        depth = depth_list(df["Depth_m"])
        phase_dict, phase_rle = dict_rle(df["Phase"])
        source_dict, source_rle = dict_rle(df["Source_Block"])

        t_valid = [v for v in t_raw if v is not None]
        payload = {
            "format_version": FORMAT_VERSION,
            "site": site,
            "source_chunk": chunk_file,
            "n": n,
            "t0": t0,
            "t_delta": t_delta,
            "dt": dt,
            "depth": depth,
            "phase_dict": phase_dict,
            "phase_rle": phase_rle,
            "source_dict": source_dict,
            "source_rle": source_rle,
            "notes": ("t values reconstruct DateTime_GMT_minus_5 as POSIXct "
                      "epoch seconds; GMT-5 wall-clock time = epoch - 18000 "
                      "seconds rendered with UTC formatting. dt is the "
                      "original character DateTime, verbatim, for QA/QC."),
        }

        out_path = os.path.join(OUT_DIR, "sites", f"{site}.json.gz")
        raw = json.dumps(payload, separators=(",", ":"),
                         ensure_ascii=False, allow_nan=False).encode("utf-8")
        with open(out_path, "wb") as fh:
            # mtime=0 keeps the output byte-identical across reruns.
            with gzip.GzipFile(fileobj=fh, mode="wb", compresslevel=9, mtime=0) as gz:
                gz.write(raw)

        xw = xwalk_by_site.get(site, {})
        d_valid = [v for v in depth if v is not None]
        manifest_sites.append({
            "id": site,
            "file": f"sites/{site}.json.gz",
            "source_chunk": chunk_file,
            "rows": n,
            "rows_missing_time": sum(1 for v in t_raw if v is None),
            "rows_missing_depth": sum(1 for v in depth if v is None),
            "t_min": min(t_valid) if t_valid else None,
            "t_max": max(t_valid) if t_valid else None,
            "depth_min": min(d_valid) if d_valid else None,
            "depth_max": max(d_valid) if d_valid else None,
            "phases": phase_dict,
            "source_blocks": source_dict,
            "site_name": xw.get("Site_Name"),
            "standard_site_code": xw.get("Standard_Site_Code"),
            "sensor_id": xw.get("Sensor_ID"),
            "bytes_gz": os.path.getsize(out_path),
        })

    manifest = {
        "format_version": FORMAT_VERSION,
        "generated_utc": datetime.now(timezone.utc).strftime("%Y-%m-%dT%H:%M:%SZ"),
        "timezone_note": ("DateTime_GMT_minus_5 is stored as POSIXct epoch "
                          "seconds; the app renders it as fixed-offset GMT-5 "
                          "wall-clock time (epoch - 18000 s, UTC-formatted). "
                          "No daylight-saving adjustment is applied."),
        "columns": ["DateTime", "DateTime_GMT_minus_5", "Depth_m", "Phase",
                    "Source_Block"],
        "sites": manifest_sites,
        "crosswalk": crosswalk,
    }
    with open(os.path.join(OUT_DIR, "manifest.json"), "w", encoding="utf-8") as fh:
        json.dump(manifest, fh, indent=1, ensure_ascii=False)

    total = sum(s["bytes_gz"] for s in manifest_sites)
    print(f"Done. {len(manifest_sites)} sites, {total/1e6:.1f} MB gzipped total.")


if __name__ == "__main__":
    main()

File: preprocess/preprocess.R
# Preprocess depth RDS chunks into browser-readable files for GitHub Pages.
#
# R equivalent of preprocess/preprocess.py (either script produces the same
# output format; run whichever toolchain you have). Reads the five
# Chunk_*.rds files (each a named list of per-site data frames) plus
# Depth_Site_Name_Crosswalk.rds and writes:
#
#   data/manifest.json          - site index, crosswalk, format documentation
#   data/sites/<SITE>.json.gz   - one gzipped columnar JSON file per site
#
# Original values are preserved exactly: DateTime verbatim, POSIXct epoch
# seconds delta-encoded losslessly, Depth_m at full precision, Phase and
# Source_Block run-length encoded.
#
# Requires: install.packages("jsonlite")
# Usage:    Rscript preprocess/preprocess.R   (from the repository root)

library(jsonlite)

chunk_files <- c(
  "Chunk_01_Blue_Ridge.rds",
  "Chunk_02_Central_Maryland.rds",
  "Chunk_03_PRV.rds",
  "Chunk_04_Roberts_Field.rds",
  "Chunk_05_Shannon_Run.rds"
)
expected_sites <- c("BR_2", "BR_3", "CM_1", "CM_2", "PRV_1", "PRV_2",
                    "RF_1", "RF_2", "SR_1", "SR_3")
format_version <- 1

dir.create("data/sites", recursive = TRUE, showWarnings = FALSE)

crosswalk <- readRDS("Depth_Site_Name_Crosswalk.rds")

rle_encode <- function(codes) {
  r <- rle(codes)
  mapply(function(v, l) list(v, l), r$values, r$lengths,
         SIMPLIFY = FALSE, USE.NAMES = FALSE)
}

encode_time <- function(t_num) {
  # Delta-encode epoch seconds; NA -> null; first non-NA stored as 0 (= t0).
  out <- vector("list", length(t_num))
  t0 <- NA_real_
  prev <- NA_real_
  for (i in seq_along(t_num)) {
    v <- t_num[i]
    if (is.na(v)) { out[[i]] <- NA; next }
    if (is.na(t0)) { t0 <- v; out[[i]] <- 0 } else out[[i]] <- v - prev
    prev <- v
  }
  list(t0 = t0, deltas = out)
}

manifest_sites <- list()

for (f in chunk_files) {
  message("Reading ", f)
  chunk <- readRDS(f)
  for (site in names(chunk)) {
    df <- chunk[[site]]
    n <- nrow(df)
    message("Encoding ", site, ": ", n, " rows")

    t_num <- as.numeric(df$DateTime_GMT_minus_5)  # POSIXct -> epoch seconds
    enc <- encode_time(t_num)

    phase_dict <- unique(as.character(df$Phase))
    source_dict <- unique(as.character(df$Source_Block))
    phase_codes <- match(as.character(df$Phase), phase_dict) - 1L
    source_codes <- match(as.character(df$Source_Block), source_dict) - 1L

    payload <- list(
      format_version = format_version,
      site = site,
      source_chunk = f,
      n = n,
      t0 = enc$t0,
      t_delta = enc$deltas,
      dt = as.character(df$DateTime),
      depth = as.list(ifelse(is.na(df$Depth_m), NA, df$Depth_m)),
      phase_dict = phase_dict,
      phase_rle = rle_encode(phase_codes),
      source_dict = source_dict,
      source_rle = rle_encode(source_codes),
      notes = paste("t values reconstruct DateTime_GMT_minus_5 as POSIXct",
                    "epoch seconds; GMT-5 wall-clock time = epoch - 18000",
                    "seconds rendered with UTC formatting. dt is the original",
                    "character DateTime, verbatim, for QA/QC.")
    )

    json <- toJSON(payload, auto_unbox = TRUE, null = "null", na = "null",
                   digits = NA)
    con <- gzfile(file.path("data", "sites", paste0(site, ".json.gz")),
                  "wb", compression = 9)
    writeBin(charToRaw(json), con)
    close(con)

    xw <- crosswalk[crosswalk$Final_List_Name == site, ]
    tv <- t_num[!is.na(t_num)]
    dv <- df$Depth_m[!is.na(df$Depth_m)]
    manifest_sites[[length(manifest_sites) + 1]] <- list(
      id = site,
      file = paste0("sites/", site, ".json.gz"),
      source_chunk = f,
      rows = n,
      rows_missing_time = sum(is.na(t_num)),
      rows_missing_depth = sum(is.na(df$Depth_m)),
      t_min = if (length(tv)) min(tv) else NA,
      t_max = if (length(tv)) max(tv) else NA,
      depth_min = if (length(dv)) min(dv) else NA,
      depth_max = if (length(dv)) max(dv) else NA,
      phases = phase_dict,
      source_blocks = source_dict,
      site_name = if (nrow(xw)) xw$Site_Name[1] else NA,
      standard_site_code = if (nrow(xw)) xw$Standard_Site_Code[1] else NA,
      sensor_id = if (nrow(xw)) xw$Sensor_ID[1] else NA,
      bytes_gz = file.size(file.path("data", "sites", paste0(site, ".json.gz")))
    )
  }
}

got <- vapply(manifest_sites, function(s) s$id, character(1))
if (!all(expected_sites %in% got)) {
  stop("Missing expected sites: ",
       paste(setdiff(expected_sites, got), collapse = ", "))
}

manifest <- list(
  format_version = format_version,
  generated_utc = format(Sys.time(), "%Y-%m-%dT%H:%M:%SZ", tz = "UTC"),
  timezone_note = paste("DateTime_GMT_minus_5 is stored as POSIXct epoch",
                        "seconds; the app renders it as fixed-offset GMT-5",
                        "wall-clock time (epoch - 18000 s, UTC-formatted).",
                        "No daylight-saving adjustment is applied."),
  columns = c("DateTime", "DateTime_GMT_minus_5", "Depth_m", "Phase",
              "Source_Block"),
  sites = manifest_sites,
  crosswalk = crosswalk
)
write(toJSON(manifest, auto_unbox = TRUE, null = "null", na = "null",
             digits = NA, pretty = TRUE), "data/manifest.json")
message("Done.")
